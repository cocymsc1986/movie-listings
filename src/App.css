

.app {
  padding: 1rem;
}

.filters__genres {
  display: flex;
  flex-flow: wrap;
}

.filters__genre {
  width: 25%;
  padding: 0.75rem;
  border: 1px solid #fff;
  background: #00bab3;
  color: white;
  text-align: center;
  box-sizing: border-box;
  cursor: pointer;
}

.filters__genre.filters__genre--selected {
  background: #019b95;
  box-shadow: inset 3px 3px 10px #018782;
}

.filters__genre label {
  pointer-events: none;
}

.filters__genre input[type=checkbox] {
  display: none;
}

.movies__results {
  display: flex;
  flex-flow: wrap;
}

.movies__result {
  width: 50%;
  display: flex;
  margin-bottom: 1rem;
}

.movies__image {
  margin-right: 0.5rem;
}

.movies__genre-labels {
  
}

.movies__genre-label {
  display: inline-block;
  padding: 0.25rem 0.5rem;
  margin-top: 0.5rem;
  margin-right: 0.5rem;
  font-size: 0.75rem;
  background: #5900bf;
  color: #fff;
}

.movies__genre-label:last-child {
  margin-right: 0;
}

@media only screen and (max-width: 720px) {
  .filters__genre {
    width: 50%;
  }

  .movies__result { 
    width: 100%;
  }
}