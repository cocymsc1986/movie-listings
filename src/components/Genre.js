import React from 'react';

const Genre = ({ name, id, onClick }) => {
    return (
        <div className="filters__genre" onClick={onClick}>
            <label htmlFor={`genre-${id}`}>{name}</label>
            <input type="checkbox" id={`genre-${id}`} value={id} />
        </div>
    );
};

export default Genre;