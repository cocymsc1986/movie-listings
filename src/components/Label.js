import React from 'react';

const Label = ({ label }) => {
	return (
		<div className="movies__genre-label">{label}</div>
	)
};

export default Label;