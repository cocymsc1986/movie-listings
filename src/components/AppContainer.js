import React, { Component } from 'react';
import { connect } from 'react-redux';
import App from './App';

import genreActions from '../actions/genres';
import movieActions from '../actions/movies';

export class AppContainer extends Component {
  constructor(props) {
    super(props);

    this.state = {
      genres: null,
      movies: null,
      genre_filter: [],
      score_filter: 3
    }

    this.clickCheckBox = this.clickCheckBox.bind(this);
    this.updateScoreFilter = this.updateScoreFilter.bind(this);
  }

  async componentWillMount() {
    await Promise.all([this.props.getGenres(), this.props.getLatest()]);
    const { results: movies } = this.props.movies.data;

    // Sort by popularity
    movies.sort((a, b) => b.popularity - a.popularity)

    // Get relevant genres
    const genres = this.filterGenres();

    this.setState({
      genres,
      movies
    });
  }

  filterGenres() {
    const { genres: { data: { genres } }, movies: { data: { results } } } = this.props;
    let applicableGenres = [];

    // Find which genres we need from the returned movies and push into array applicableGenres
    results.forEach(movie => {
      movie.genre_ids.forEach(id => {
        if (!applicableGenres.includes(id)) { 
          applicableGenres.push(id)
        }
      });
    });

    // filter the genre list to only include what is in applicableGenres
    return genres.filter(genre => applicableGenres.includes(genre.id));
  }

  updateScoreFilter(event) {
    this.setState({
      score_filter: event.target.value
    });
  };

  clickCheckBox(event) {
    const checkbox = event.target.children[1];
    const filters = this.state.genre_filter;

    checkbox.checked = !checkbox.checked;

    if (checkbox.checked) {
      filters.push(checkbox.value);
      event.target.classList.add('filters__genre--selected');
    } else {
      const index = filters.indexOf(checkbox.value);
      filters.splice(index, 1);
      event.target.classList.remove('filters__genre--selected');
    };

    this.setState({
      genre_filter: filters
    });
  }

  filterMovies(movies) {
    const { genre_filter, score_filter } = this.state;

    return movies.filter(movie => {
      return genre_filter.every(id => movie.genre_ids.includes(parseInt(id, 10))) && movie.vote_average >= score_filter;
    });
  }

  render() {
    const { genres, movies, score_filter } = this.state;
    const { genres: { loading: genresLoading, error: genresError }, movies: { loading: moviesLoading, error: moviesError } } = this.props;

    return (
      <App
        genres={genres}
        movies={movies}
        score_filter={score_filter}
        genresLoading={genresLoading}
        moviesLoading={moviesLoading}
        filterMovies={() => this.filterMovies(movies)}
        clickCheckBox={(e) => this.clickCheckBox(e)}
        updateScoreFilter={(e) => this.updateScoreFilter(e)}
        genresError={genresError}
        moviesError={moviesError}
      />
    )
  }
}

const mapStateToProps = ({ genres, movies }) => ({
  genres,
  movies
});

export default connect(
  mapStateToProps,
  { ...genreActions, ...movieActions }
)(AppContainer);