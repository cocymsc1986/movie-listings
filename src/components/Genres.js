import React, { Fragment } from 'react';

import Genre from './Genre';

const Genres = ({ genres, clickCheckBox }) => {
    return (
        <Fragment>
            {genres && genres.map((genre, key) => {
                return (
                    <Genre
                        key={key}
                        onClick={clickCheckBox}
                        name={genre.name}
                        id={genre.id}
                    />
                );
            })}
        </Fragment>
    );
};

export default Genres;