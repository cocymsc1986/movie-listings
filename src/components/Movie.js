import React, { Component } from 'react';
import Label from './Label';

export class Movie extends Component {

	render() {
		const { genreLabels, imagePath, title } = this.props;

		return (
			<article className="movies__result">
				<div className="movies__image"><img src={`https://image.tmdb.org/t/p/w200/${imagePath}`} alt={`${title}`} /></div>
				<div>
					<div className="movies__title"><h3>{title}</h3></div>
					<div className="movies__genre-labels">{genreLabels.map((label, key) => <Label label={label} key={key} />)}</div>
				</div>
			</article>
		);
	}
}

export default Movie;