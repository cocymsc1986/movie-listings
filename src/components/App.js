import React, { Component } from 'react';

import Movie from './MovieContainer';
import Rating from './Rating'
import '../App.css';

class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      genres: null,
      movies: null,
      genre_filter: [],
      score_filter: 0
    }

    this.clickCheckBox = this.clickCheckBox.bind(this);
    this.updateScoreFilter = this.updateScoreFilter.bind(this);
  }

  /* Move functions into App Container - make stateless */
  async componentWillMount() {
    await Promise.all([this.props.getGenres(), this.props.getLatest()]);
    const { genres } = this.props.genres.data;
    const { results: movies } = this.props.movies.data;

    this.setState({
      genres,
      movies
    });
  }

  updateScoreFilter(event) {
    this.setState({
      score_filter: event.target.value
    });
  };

  clickCheckBox(event) {
    const checkbox = event.target.children[1];
    const filters = this.state.genre_filter;

    checkbox.checked = !checkbox.checked;

    if (checkbox.checked) {
      filters.push(checkbox.value);
      event.target.classList.add('filters__genre--selected');
    } else {
      const index = filters.indexOf(checkbox.value);
      filters.splice(index, 1);
      event.target.classList.remove('filters__genre--selected');
    };

    this.setState({
      genre_filter: filters
    });
  }

  filterMovies(movies) {
    const { genre_filter, score_filter } = this.state;

    return movies.filter(movie => {
      return genre_filter.every(id => movie.genre_ids.includes(parseInt(id, 10))) && movie.vote_average > score_filter;
    });
  }

  render() {
    const { genres, movies, score_filter } = this.state;

    return (
      <main className="app">
        <section className="filters">
          <h2>Filters</h2>
          <Rating 
            ratingValue={score_filter}
            updateScoreFilter={this.updateScoreFilter}
          />
          {/* Create Genres component */}
          <h3>Genres</h3>
          <div className="filters__genres">
            {genres && genres.map((genre, key) => {
              return (
                <div className="filters__genre" onClick={this.clickCheckBox} key={key}>
                  <label htmlFor={`genre-${genre.id}`}>{genre.name}</label>
                  <input type="checkbox" id={`genre-${genre.id}`} value={genre.id} />
                </div>
              )
            })}
          </div>
        </section>

        <section class="movies"> 
          <h2>Movies</h2>
          <div className="movies__results">
            {movies && this.filterMovies(movies).map((movie, key) => {
              return (
                <Movie
                  imagePath={movie.poster_path}
                  title={movie.title}
                  genreIds={movie.genre_ids}
                  genres={genres}
                  key={key}
                />
              )
            })}
          </div>
        </section>
      </main>
    );
  }
}

export default App;
