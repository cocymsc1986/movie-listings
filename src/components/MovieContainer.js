import React from 'react';
import Movie from './Movie';

export const getGenreLabels = (genreIds, genres) => {
	return genreIds.map(id => genres.find(genre => genre.id === id).name);
};

export const MovieContainer = ({ imagePath, title, genreIds, genres }) => {
	return (
		<Movie
			imagePath={imagePath}
			title={title}
			genreIds={genreIds}
			genreLabels={getGenreLabels(genreIds, genres)}
		/>
	);
}

export default MovieContainer;