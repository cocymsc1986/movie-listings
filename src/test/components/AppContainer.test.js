import React from 'react';

import { AppContainer } from '../../components/AppContainer';
import App from '../../components/App';

describe('AppContainer component', () => {
	let wrapper;
	const mockGetGenres = jest.fn();
	const mockGetLatest = jest.fn();

	const props = {
		getGenres: mockGetGenres,
		getLatest: mockGetLatest,
		genres: {
			data: {
				genres: [
					{
						id: 1,
						name: 'Action'
					},
					{
						id: 2,
						name: 'Adventure'
					},
					{
						id: 3,
						name: 'Comedy'
					},
					{
						id: 4,
						name: 'Romance'
					}
				],
			},
			loading: false
		},
		movies: {
			data: {
				results: [
					{
						name: 'test-genres',
						popularity: 5,
						genre_ids: [2, 3]
					},
					{
						name: 'test-genres',
						popularity: 5,
						genre_ids: [3, 4]
					}
				]
			},
			loading: false
		}
	}

	beforeEach(() => {
		wrapper = shallow(<AppContainer {...props} />);
	});

	test('On component mount should call actions', () => {
		expect(mockGetGenres).toHaveBeenCalled();
		expect(mockGetLatest).toHaveBeenCalled();
	});

	test('updateScoreFilter should update state', () => {
		wrapper.instance().updateScoreFilter({ target: { value: 1 } } );
		expect(wrapper.state().score_filter).toBe(1);
	});

	test('filterGenres should return filtered genre list', () => {
		wrapper.instance().filterGenres();
		expect(wrapper.state().genres).toEqual([
			{
				id: 2,
				name: 'Adventure'
			},
			{
				id: 3,
				name: 'Comedy'
			},
			{
				id: 4,
				name: 'Romance'
			}
		]);
	});

	test('clickCheckBox should update state', () => {
		wrapper.instance().clickCheckBox({ target: { children: [{}, { value: 1, checked: false }], classList: { add: () => {}} } } );
		expect(wrapper.state().genre_filter).toEqual([1]);
	});

	test('filterMovies should return filtered movies from state filters', () => {
		const movies = [
			{
				genre_ids: [12, 14, 18, 24],
				vote_average: 8.1,
				name: 'Armageddon'
			},
			{
				genre_ids: [14, 24],
				vote_average: 9,
				name: 'Blues Brothers'
			},
			{
				genre_ids: [12, 14, 29],
				vote_average: 6,
				name: 'Avengers Assemble'
			},
		];

		wrapper.setState({
			genre_filter: [12, 14],
			score_filter: 7
		});

		const result = wrapper.instance().filterMovies(movies);

		expect(result).toEqual([
			{
				genre_ids: [12, 14, 18, 24],
				vote_average: 8.1,
				name: 'Armageddon'
			}
		]);

	});

	test('Should render <App /> with correct props', () => {
		wrapper.setState({
			genres: 'test-genres',
			movies: 'test-movies',
			score_filter: 1
		});

		expect(wrapper.find(App).prop('genres')).toBe('test-genres');
		expect(wrapper.find(App).prop('movies')).toBe('test-movies');
		expect(wrapper.find(App).prop('score_filter')).toBe(1);
		expect(wrapper.find(App).prop('genresLoading')).toBe(false);
		expect(wrapper.find(App).prop('moviesLoading')).toBe(false);
	});

	test('Should render', () => {
		expect(wrapper).toMatchSnapshot();
	});
});