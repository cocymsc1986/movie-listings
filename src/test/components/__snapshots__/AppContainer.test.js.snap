// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppContainer component Should render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AppContainer
    genres={
        Object {
            "data": Object {
              "genres": Array [
                Object {
                  "id": 1,
                  "name": "Action",
                },
                Object {
                  "id": 2,
                  "name": "Adventure",
                },
                Object {
                  "id": 3,
                  "name": "Comedy",
                },
                Object {
                  "id": 4,
                  "name": "Romance",
                },
              ],
            },
            "loading": false,
          }
    }
    getGenres={[Function]}
    getLatest={[Function]}
    movies={
        Object {
            "data": Object {
              "results": Array [
                Object {
                  "genre_ids": Array [
                    2,
                    3,
                  ],
                  "name": "test-genres",
                  "popularity": 5,
                },
                Object {
                  "genre_ids": Array [
                    3,
                    4,
                  ],
                  "name": "test-genres",
                  "popularity": 5,
                },
              ],
            },
            "loading": false,
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "clickCheckBox": [Function],
      "filterMovies": [Function],
      "genres": null,
      "genresError": undefined,
      "genresLoading": false,
      "movies": null,
      "moviesError": undefined,
      "moviesLoading": false,
      "score_filter": 0,
      "updateScoreFilter": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "clickCheckBox": [Function],
        "filterMovies": [Function],
        "genres": null,
        "genresError": undefined,
        "genresLoading": false,
        "movies": null,
        "moviesError": undefined,
        "moviesLoading": false,
        "score_filter": 0,
        "updateScoreFilter": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
